File: src/modules/minigames/lib/board.js
// A generic grid + line finder, ported from `minigames/board.py` in
// Brandjuh/FireAndRescueAcademyCogs (M26.2, owner: "Vervang deze met …").
//
// The cog's own comment says these functions were "originally handwritten in
// C# by me, and converted to Python with an LLM" — so this is the third
// translation. The scan ORDER is preserved exactly (horizontals, verticals,
// then both diagonal families) because `findLines` reports the FIRST line it
// completes, and a reordered scan would highlight a different set of four on
// a board that happens to contain two.
//
// Pure: plain numbers in, plain numbers out. No discord.js.

/** A width × height grid stored as one flat array, like the source. */
export class Board {
  constructor(width, height, fill = null) {
    this.width = width;
    this.height = height;
    this.data = new Array(width * height).fill(fill);
  }

  #index(x, y) {
    if (!(x >= 0 && x < this.width && y >= 0 && y < this.height)) {
      throw new RangeError('Board index out of range');
    }
    return y * this.width + x;
  }

  get(x, y) {
    return this.data[this.#index(x, y)];
  }

  set(x, y, value) {
    this.data[this.#index(x, y)] = value;
  }

  copy() {
    const next = new Board(this.width, this.height);
    next.data = [...this.data];
    return next;
  }
}

/**
 * Is there a run of `length` cells holding `value`? Optionally collect it.
 *
 * The source mutates a shared `line` list from a closure and resets it between
 * scan directions; that is reproduced rather than rewritten, because the exact
 * reset points decide which cells land in `result` when a run is longer than
 * `length` (the source appends only the overhanging cell in that case).
 */
export function findLines(board, value, length, result = null) {
  let win = false;
  let line = [];

  const checkCell = (x, y) => {
    if (board.get(x, y) === value) {
      line.push([x, y]);
      if (line.length >= length) {
        win = true;
        if (result) {
          if (line.length === length) result.push(...line);
          else result.push([x, y]);
        }
      }
    } else {
      line = [];
    }
  };

  for (let y = 0; y < board.height; y += 1) {
    line = [];
    for (let x = 0; x < board.width; x += 1) checkCell(x, y);
  }
  for (let x = 0; x < board.width; x += 1) {
    line = [];
    for (let y = 0; y < board.height; y += 1) checkCell(x, y);
  }
  // Top-left → bottom-right.
  for (let d = length - 1; d < board.width + board.height - length + 1; d += 1) {
    line = [];
    for (let y = 0; y < board.height; y += 1) {
      const x = d - y;
      if (x >= 0 && x < board.width) checkCell(x, y);
    }
  }
  // Top-right → bottom-left.
  for (let d = length - 1; d < board.width + board.height - length + 1; d += 1) {
    line = [];
    for (let y = 0; y < board.height; y += 1) {
      const x = board.width - 1 - d + y;
      if (x >= 0 && x < board.width) checkCell(x, y);
    }
  }
  return win;
}

/**
 * The one empty cell that would complete a line for `value`, or null.
 *
 * Used by Tic-Tac-Toe's opponent. Note the source's quirk, kept deliberately:
 * `count` and `missing` are not reset when a run is broken by an ENEMY piece,
 * so a row like `X X O _` can still report the gap. It is a weaker player than
 * a correct implementation would be — which is the point, since the owner
 * chose this opponent over ours knowing it is the weaker one.
 */
export function tryCompleteLine(board, value, empty, length) {
  const scan = (cells) => {
    let count = 0;
    let missing = null;
    for (const [x, y] of cells) {
      const cell = board.get(x, y);
      if (cell === value) count += 1;
      else if (cell === empty) missing = [x, y];
      if (count === length - 1 && missing !== null) return missing;
    }
    return null;
  };

  for (let y = 0; y < board.height; y += 1) {
    const hit = scan([...Array(board.width).keys()].map((x) => [x, y]));
    if (hit) return hit;
  }
  for (let x = 0; x < board.width; x += 1) {
    const hit = scan([...Array(board.height).keys()].map((y) => [x, y]));
    if (hit) return hit;
  }
  for (let d = length - 1; d < board.width + board.height - length + 1; d += 1) {
    const cells = [];
    for (let y = 0; y < board.height; y += 1) {
      const x = d - y;
      if (x >= 0 && x < board.width) cells.push([x, y]);
    }
    const hit = scan(cells);
    if (hit) return hit;
  }
  for (let d = length - 1; d < board.width + board.height - length + 1; d += 1) {
    const cells = [];
    for (let y = 0; y < board.height; y += 1) {
      const x = board.width - 1 - d + y;
      if (x >= 0 && x < board.width) cells.push([x, y]);
    }
    const hit = scan(cells);
    if (hit) return hit;
  }
  return null;
}

File: cuffminigames/board.py
"""Board primitives, copied verbatim from crab-cogs `minigames/board.py`.

The only addition is `winning_lines` at the bottom — see its docstring for why
`find_lines`' own `result` parameter could not answer the question this cog
needs answered.
"""

from typing import Any, List, Tuple, Optional


Pos = Tuple[int, int]


class Board:
    def __init__(self, width: int, height: int, fill: Any = None):
        self.width = width
        self.height = height
        self._data = [fill] * (width * height)

    def _index(self, x: int, y: int):
        if not (0 <= x < self.width and 0 <= y < self.height):
            raise IndexError("Board index out of range")
        return y * self.width + x

    def __getitem__(self, pos: Pos):
        x, y = pos
        return self._data[self._index(x, y)]

    def __setitem__(self, pos: Pos, value: Any):
        x, y = pos
        self._data[self._index(x, y)] = value

    def copy(self):
        new_board = Board(self.width, self.height)
        new_board._data = list(self._data)
        return new_board


# These functions were originally handwritten in C# by me, and converted to Python with an LLM


def find_lines(board: Board, value: Any, length: int, result: Optional[List[Pos]] = None):
    win = False
    line: List[Pos] = []

    def check_cell(x: int, y: int):
        nonlocal win, line
        pos = (x, y)
        if board[pos] == value:
            line.append(pos)
            if len(line) >= length:
                win = True
                if result is not None:
                    if len(line) == length:
                        result.extend(line)
                    else:
                        result.append(pos)
        else:
            line = []

    # Horizontals
    for y in range(board.height):
        line = []
        for x in range(board.width):
            check_cell(x, y)

    # Verticals
    for x in range(board.width):
        line = []
        for y in range(board.height):
            check_cell(x, y)

    # Diagonals (top-left to bottom-right)
    for d in range(length - 1, board.width + board.height - length + 1):
        line = []
        for y in range(board.height):
            x = d - y
            if 0 <= x < board.width:
                check_cell(x, y)

    # Diagonals (top-right to bottom-left)
    for d in range(length - 1, board.width + board.height - length + 1):
        line = []
        for y in range(board.height):
            x = board.width - 1 - d + y
            if 0 <= x < board.width:
                check_cell(x, y)

    return win


def try_complete_line(board: Board, value: Any, empty: Any, length: int) -> Optional[Pos]:
    # Horizontals
    for y in range(board.height):
        count, missing = 0, None
        for x in range(board.width):
            if board[x, y] == value:
                count += 1
            elif board[x, y] is empty:
                missing = (x, y)
            if count == length - 1 and missing is not None:
                return missing

    # Verticals
    for x in range(board.width):
        count, missing = 0, None
        for y in range(board.height):
            if board[x, y] == value:
                count += 1
            elif board[x, y] is empty:
                missing = (x, y)
            if count == length - 1 and missing is not None:
                return missing

    # Diagonals (top-left to bottom-right)
    for d in range(length - 1, board.width + board.height - length + 1):
        count, missing = 0, None
        for y in range(board.height):
            x = d - y
            if 0 <= x < board.width:
                if board[x, y] == value:
                    count += 1
                elif board[x, y] is empty:
                    missing = (x, y)
                if count == length - 1 and missing is not None:
                    return missing

    # Diagonals (top-right to bottom-left)
    for d in range(length - 1, board.width + board.height - length + 1):
        count, missing = 0, None
        for y in range(board.height):
            x = board.width - 1 - d + y
            if 0 <= x < board.width:
                if board[x, y] == value:
                    count += 1
                elif board[x, y] is empty:
                    missing = (x, y)
                if count == length - 1 and missing is not None:
                    return missing

    return None


#: The four straight directions a line can run in. `(1, -1)` is the up-right
#: diagonal; its mirror `(-1, 1)` is the same line walked backwards, so scanning
#: only these four covers every line exactly once.
DIRECTIONS: List[Pos] = [(1, 0), (0, 1), (1, 1), (1, -1)]


def winning_lines(board: Board, value: Any, length: int) -> List[List[Pos]]:
    """Every straight run of at least `length` cells holding `value`.

    `find_lines` above answers *"is there a win"*, and its optional `result`
    parameter flattens the cells of every direction into one list — so a player
    who completes two lines at once produces a blob that cannot be told apart
    again. Highlighting a board and naming the line both need the runs kept
    separate, which is what this returns.

    A run is only started where the cell *behind* it breaks the run, so each
    line is emitted once rather than once per starting offset.
    """
    lines: List[List[Pos]] = []
    for dx, dy in DIRECTIONS:
        for y in range(board.height):
            for x in range(board.width):
                px, py = x - dx, y - dy
                if 0 <= px < board.width and 0 <= py < board.height and board[px, py] == value:
                    continue  # not the start of this run
                run: List[Pos] = []
                cx, cy = x, y
                while 0 <= cx < board.width and 0 <= cy < board.height and board[cx, cy] == value:
                    run.append((cx, cy))
                    cx, cy = cx + dx, cy + dy
                if len(run) >= length:
                    lines.append(run)
    return lines
